### PR TITLE
facets: new filter for ranges

### PR DESCRIPTION
**Pull request by nikofil**
* NEW Implements a filter that filters a field based on a range of
 values. This is used in the year filter facet for invenio-search-ui.

Signed-off-by: Nikos Filippakis <nikolaos.filippakis@cern.ch>

### CODE REVIEW

File: invenio_records_rest/facets.py [hunk 1]
```diff
@@ -38,6 +38,27 @@ def inner(values):
     return inner
 
 
+def range_filter(field, **kwargs):
+    """Create a range filter."""
+    def inner(values):
+        range_ends = [int(val) for val in values[0].split(',')]
```
Comment (jirikuncar) [inline invenio_records_rest/facets.py]: - [ ] We should be able to support not only `int`.
- [ ] Why using `,` as separator for ranges?
- [ ] Have you considered using Elasticsearch-DSL `Range` instead?

File: invenio_records_rest/facets.py [hunk 2]
```diff
@@ -38,6 +39,42 @@ def inner(values):
     return inner
 
 
+def range_filter(field, **kwargs):
```
Comment (lnielsen) [inline invenio_records_rest/facets.py]:
```python
def range_filter(field, start_date_math=None, end_date_math=None, **kwargs):
    """Create a range filter."""
    def inner(values):
        start, end = values[0].split('--')

        if start_date_math:
            start = '{0}||{1}'.format(start, start_date_math)

        if end_date_math:
            end = '{0}||{1}'.format(end, end_date_math)

        kwargs.update({
            'gte': start,
            'lte': end
        })

        return Range(**{field: params})

    return inner
```

File: invenio_records_rest/facets.py [hunk 3]
```diff
@@ -38,6 +39,42 @@ def inner(values):
...
+    if 'start_date_math' in kwargs:
+        start_date_math = '||' + args['start_date_math']
+        del args['start_date_math']
+
+    if 'end_date_math' in kwargs:
+        end_date_math = '||' + args['end_date_math']
+        del args['end_date_math']
+
+    def inner(values):
+        [start, end] = values[0].split('--')
```
Comment (lnielsen) [inline invenio_records_rest/facets.py]: What happens on bad input data - e.g. ``1940`` or ``1940-``

Comment (nikofil) [inline invenio_records_rest/facets.py]: If no '--' is found, the terms_filter will be now used instead, as I believe 1940- may be a valid value in some context.

File: invenio_records_rest/facets.py (hunk 2 above)
Comment (jirikuncar) [inline invenio_records_rest/facets.py]: `s/return Range(**{field: params})/return Range(**{field: kwargs})/`?

**Comment by lnielsen** [comment]
The ES [range query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html) falls in two types:

- TermRangeQuery (strings)
- NumericRangeQuery (dates + numbers)

Both types needs one or more of the following parameters: ``gt``, ``gte``, ``lt``, ``lte``

The REST API can be queried like this:

- ``/api/records/?test=<value1>&test=<value2>``

With will result in *one* call to the ``inner`` method with ``values`` being a list ``[<value 1>, <value 2>]``

Issues:
- Handling multiple input values -> should produce multiple range queries or perhaps fail.
- Parsing and validating of one input value, and mapping this into one or more of the parameters (``gt``, ``gte``, ``lt``, ``lte``). 

Regarding the parsing we could do something along the lines of:
- ``1940--2009`` -> ``gte=1940, lte=2009``
- ``--2009`` -> ``lte=2009``
- ``1940--`` -> ``gte=1940``
- ``1940`` -> ``gte=1940``
- ``>1940--<2009`` -> ``gt=1940, lt=2009``
- ``--<2009`` -> ``lt=2009``
- ``>1940--`` -> ``gt=1940``
- ``a--m`` -> ``gte=a, lte=m``
- ...

In case the parsing fails, I think we should raise a ``invenio_rest.errors.RESTValidationError``:

```python
raise RESTValidationError(errors=[FieldError('test', 'Invalid range format.')])
```

This will result in a 400 HTTP error, with a message staying that the format is invalid.

**Comment by nikofil** [comment]
In which case would the parsing fail? Only case I can think of would be just `--`.
Besides that, I don't see how multiple values would fit into the range query, unless we did something like `term=gt2016&term=lte2020` instead of `2016--<2020`. Of course, I have little experience with ElasticSearch, so any suggestions are welcome.

File: tests/test_facets.py [hunk 4]
```diff
@@ -42,6 +45,26 @@ def test_terms_filter():
...
+    assert f(['1821--1940']) == Range(test={
+        'gte': '1821||startmath', 'lte': '1940||endmath',
+    })
+    assert f(['>1821--']) == Range(test={'gt': '1821||startmath'})
+    assert f(['1821--<1940']) == Range(test={'gte': '1821||startmath',
+                                             'lt': '1940||endmath'})
+
+    exception_raised = False
+    try:
+        f(['2016'])
```
Comment (lnielsen) [inline tests/test_facets.py]:
```python
assert pytest.raises(RESTValidationError, f, ['2016'])
```

**Pull request merged**